pkg/ring/kv/multi.go

```diff
@@ -0,0 +1,356 @@
+func (cfg *MultiConfig) RegisterFlagsWithPrefix(f *flag.FlagSet, prefix string) {
+	f.StringVar(&cfg.Primary, prefix+"multi.primary", "", "Primary backend storage used by multi-client.")
+	f.StringVar(&cfg.Secondary, prefix+"multi.secondary", "", "Secondary backend storage used by multi-client.")
+	f.BoolVar(&cfg.MirrorEnabled, prefix+"multi.mirror-enabled", false, "Mirror writes to secondary store.")
+	f.DurationVar(&cfg.MirrorTimeout, prefix+"multi.mirror-timeout", 2*time.Second, "Timeout for storing value to secondary store.")
+}
+
+// MultiRuntimeConfig has values that can change in runtime (via overrides)
+type MultiRuntimeConfig struct {
+	// Primary store used by MultiClient. Can be updated in runtime to switch to a different store (eg. consul -> etcd,
+	// or to gossip). Doing this allows nice migration between stores. Empty values are ignored.
+	PrimaryStore string `yaml:"primary"`
+
+	// Mirroring enabled or not. Nil = no change.
+	Mirroring *bool `yaml:"mirror_enabled"`
+}
+
+type kvclient struct {
+	client Client
+	name   string
+}
+
+type clientInProgress struct {
+	client int
+	cancel context.CancelFunc
+}
+
+// MultiClient implements kv.Client by forwarding all API calls to primary client.
+// Writes performed via CAS method are also (optionally) forwarded to secondary clients.
+type MultiClient struct {
+	// Available KV clients
+	clients []kvclient
+
+	mirrorTimeout    time.Duration
+	mirroringEnabled *atomic.Bool
+
+	// The primary client used for interaction.
+	primaryID *atomic.Int32
+
+	cancel context.CancelFunc
+
+	inProgressMu sync.Mutex
+	// Cancel functions for ongoing operations. key is a value from inProgressCnt.
+	// What we really need is a []context.CancelFunc, but functions cannot be compared against each other using ==,
+	// so we use this map instead.
+	inProgress    map[int]clientInProgress
+	inProgressCnt int
+}
+
+// NewMultiClient creates new MultiClient with given KV Clients.
+// First client in the slice is the primary client.
+func NewMultiClient(cfg MultiConfig, clients []kvclient) *MultiClient {
+	c := &MultiClient{
+		clients:    clients,
+		primaryID:  atomic.NewInt32(0),
+		inProgress: map[int]clientInProgress{},
+
+		mirrorTimeout:    cfg.MirrorTimeout,
+		mirroringEnabled: atomic.NewBool(cfg.MirrorEnabled),
+	}
+
+	ctx, cancelFn := context.WithCancel(context.Background())
+	c.cancel = cancelFn
+
+	if cfg.ConfigProvider != nil {
+		go c.watchConfigChannel(ctx, cfg.ConfigProvider())
+	}
+
+	c.updatePrimaryStoreGauge()
+	c.updateMirrorEnabledGauge()
+	return c
+}
+
+func (m *MultiClient) watchConfigChannel(ctx context.Context, configChannel <-chan MultiRuntimeConfig) {
+	for {
+		select {
+		case cfg, ok := <-configChannel:
+			if !ok {
+				return
+			}
+
+			if cfg.Mirroring != nil {
+				enabled := *cfg.Mirroring
+				old := m.mirroringEnabled.Swap(enabled)
+				if old != enabled {
+					level.Info(util.Logger).Log("msg", "toggled mirroring", "enabled", enabled)
```

Comment:
Can we be more descriptive in the log here. Maybe add the component (multi-kv)?

Reply:
I've added logger with component=multikv that is used for all logging from this file.